bugfix + 0.4.2-1 upgrade

jmonteer opened the pull request.

jmonteer committed:
fix: main fix

jmonteer committed:
feat: last changes + test working again

jmonteer committed:
fix: black formatting

jmonteer committed:
fix: add files

jmonteer committed:
Merge pull request #10 from jmonteer/fix/accounting

Fix/accounting

jmonteer committed:
fix: formatting

storming0x commented on contracts/Strategy.sol:
```diff
@@ -557,6 +559,31 @@ contract Strategy is BaseStrategy {
+        returns (uint256 _amountFreed)
+    {
+        // NOTE: we try to unlock all of the collateral in the strategy (which should be == totalDebt)
+        (_amountFreed, ) = liquidatePosition(balanceOfWant());
```
Should this not be instead:

`liquidatePosition(estimatedTotalAssets())`;  ??

or 
```
uint256 debtOutstanding = vault.debtOutstanding();
liquidatePosition(debtOutstanding);
```

if setEmergencyExit is called to go thru this path it will miss the debt part of the assets and report that as a loss unless im missing smth?

Main question is why in emergency state is the liquidation not accounting for all debt like in the normal path?

storming0x replied:
Maybe this should be 

```
uint256 debtOutstanding = vault.debtOutstanding();
liquidatePosition(debtOutstanding);
```

Like in the normal path?

jmonteer replied:
it is a good comment.

staked SNX is kept in the users' wallet (strategy's wallet in this case), but locked against transfers. so we are basically asking to unlock full SNX balance, which is the maximum amount we will be able to retrieve. for that, the strategy will automatically repay debt (issued sUSD)

storming0x commented:
Also a test with emergencyExit scenario seems to be failing

jmonteer committed:
feat: new hedging vault

jmonteer committed:
fix: edge case with escrowed want

jmonteer committed:
fix: github workflow

jmonteer committed:
fix: @storming0x comment

jmonteer committed:
fix: change network

jmonteer committed:
fix: edge case for event

poolpitako reviewed (approved):
LGTM

jmonteer committed:
fix: better way to solve edge case

storming0x commented on contracts/Strategy.sol:
```diff
@@ -289,6 +290,12 @@ contract Strategy is BaseStrategy {
         // NOTE: _newCollateral will always be < _lockedCollateral() so _targetDebt will always be < _currentDebt
         uint256 _amountToRepay = _currentDebt.sub(_targetDebt);
 
+        // TODO: add threshold?
```
is the TODO no longer needed?

jmonteer committed:
fix: even better way

storming0x commented:
Should wait for CI to finished to confirm we are good, overall Fix LGTM

storming0x reviewed (approved):
Fix LGTM, if CI works for tests should be good on my end

jmonteer committed:
fix: remove estimatedProfit from estimatedTotalAssets

jmonteer committed:
feat: add healthcheck

jmonteer committed:
feat: add delegatedAssets

jmonteer committed:
fix: add maxLoss to vault withdrawal

jmonteer committed:
fix: add comment

jmonteer committed:
feat: add maxLoss to vault withdrawal

jmonteer committed:
feat: add test_lossy_vault

poolpitako committed:
fix: test

poolpitako committed:
fix: black format

poolpitako committed:
fix: test fix

poolpitako committed:
fix: removed clean

jmonteer commented:
changed since LGTM: 

- added maxLoss parameter to avoid getting locked in sUSD vault if there is a lossy strategy
- added delegatedAssets method
- removed unrealised profits from estimatedTotalAssets to avoid reporting false losses

and added tests

storming0x commented:
LGTM with latest changes adding maxLoss setting, i think all vaults that handle delegation to vaults should have it

storming0x reviewed (approved):
LGTM with latest maxloss changes

poolpitako commented:
@jmonteer squash and merge.
